Fix handling of role names for grant/revoke

amrutagokhale opened the pull request:
Hive metastore API requires role names to be in lowercase. So the role names need to be converted before they are passed to the metastore API method.

Added product test for PUBLIC role. Also corrected GRANT/REVOKE documentation for PUBLIC role.

Fixes https://github.com/prestodb/presto/issues/6688.

TD internal review: https://github.com/Teradata/presto/pull/431

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -769,6 +736,41 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
+                                                     String databaseName,
+                                                     String tableName,
+                                                     String grantee,
+                                                     Set<PrivilegeGrantInfo> privilegeGrantInfos)
```
formatting

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -769,6 +736,41 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
+        try {
+            if (metastoreClient.getRoleNames().contains(grantee.toLowerCase())) {
+                principalType = ROLE;
+                principalName = grantee.toLowerCase(); // Hive metastore API requires role names to be in lowercase.
```
when using case changing methods always declare the locale as `US` otherwise you get the locale of the VM which might do strange things

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -769,6 +736,41 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
+            }
+            return new PrivilegeBag(privilegeBagBuilder.build());
+        }
+        catch (TException e) {
```
I think you want to remove the exception handling from this method, so the retry handler can process the raw exception

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -769,6 +736,41 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
+        String principalName;
+
+        try {
+            if (metastoreClient.getRoleNames().contains(grantee.toLowerCase())) {
```
If the grantee is `PUBLIC`, can we skip the call to load the roles since it is a reserved name?

amrutagokhale commented:
@dain  Addressed your comments.

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -734,26 +719,9 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
-                            }
-                            // TODO: Check whether the user/role exists in the hive metastore.
-                            metastoreClient.revokePrivileges(new PrivilegeBag(privilegeBagBuilder.build()));
+                            metastoreClient.revokePrivileges(
```
put this on one line

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -681,25 +682,9 @@ else if (existingPrivilege.isContainedIn(requestedPrivilege)) {
-                            }
-                            // TODO: Check whether the user/role exists in the hive metastore.
-                            metastoreClient.grantPrivileges(new PrivilegeBag(privilegeBagBuilder.build()));
+                            metastoreClient.grantPrivileges(
```
put this on one line

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -769,6 +737,38 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
+                                           String databaseName,
+                                           String tableName,
+                                           String grantee,
+                                           Set<PrivilegeGrantInfo> privilegeGrantInfos)
```
formatting is still off... the arguments should be indented twice, and the first argument should be on a new line

dain commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -769,6 +737,38 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
+                                           Set<PrivilegeGrantInfo> privilegeGrantInfos)
+            throws TException
+    {
+        final String PUBLIC_ROLE_NAME = "public";
```
this is causing the build to fail since is it is not an allowed variable name.  This should be a class level constant `public static final String PUBLIC_ROLE_NAME = "PUBLIC";` declared at the top of the class

amrutagokhale commented:
@dain Addressed your comments. 

However, I found out that the code doesn't work for other role names. The problem is, the grant/revoke code is calling the existing `getPrivileges()` method to get privileges for both, users and roles. I believe `getPrivileges()` was written to get privileges only for users and not for roles (it calls `getPrivilegeSet()` which I suppose gives privileges only for users). I am going to change the `getPrivileges()` method so that it handles roles differently and will call metastore's `list_privileges` for roles. Please let me know if you have other suggestions.

dain commented:
Sounds reasonable to me

amrutagokhale commented:
@dain I made the necessary changes. Please have a look.

cawallin commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestGrantRevoke.java:
```diff
@@ -37,10 +37,16 @@
+     * (3) hive-site.xml should have the following property added:
+     *     <property>
+     *       <name>hive.users.in.admin.role</name>
+     *       <value>hive</value>
```
Is this already set in the docker images, or do you need to release them again?

amrutagokhale replied:
This property is already present in docker images. I added a comment, just to make the pre-requisites clear.

amrutagokhale commented:
I just realized that with the changes I made, role names will always take precedence over user names. For ex., if there is a role `admin` (a default role created in hive) and there is a user by the same name `admin`, `getPrivileges` method will return privileges for the role `admin` rather than the user `admin`. This may lead to unexpected results. So I guess I will have to explicitly distinguish between users and roles, passing that as an argument.

amrutagokhale commented:
@dain I am done with my changes, if you want to start reviewing.

cawallin commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -769,25 +738,103 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
-                            PrincipalPrivilegeSet privilegeSet = client.getPrivilegeSet(objectReference, user, null);
+
+                            String principalName = hivePrincipal.getPrincipalName();
+                            PrincipalPrivilegeSet privilegeSet = client.getPrivilegeSet(objectReference, principalName, null);
```
Does this also retrieve the privileges for any roles that the user is in?

amrutagokhale replied:
Yes, it does.

cawallin commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -769,25 +738,103 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
         }
     }
 
+    // GRANT and REVOKE prioritize roles over user names, given a string which is both, a role name and a user.
```
I would expect that GRANT/REVOKE would prioritize user names over roles, but you should check the spec and how other databases behave.

cawallin commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -769,25 +738,103 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
     }
 
-    private Set<HivePrivilegeInfo> getPrivileges(String user, HiveObjectRef objectReference)
+    private Set<HivePrivilegeInfo> getTablePrivileges(HivePrincipal hivePrincipal, String databaseName, String tableName)
```
It's a little confusing that this has almost the same type signature as `public Set<HivePrivilegeInfo> getTablePrivileges(String user, String databaseName, String tableName)`

amrutagokhale replied:
Yes, it is. This one is a private method where as the other is a public method. I thought about renaming the public method to `getUserTablePrivileges` since it only retrieves privileges of a user. But a better solution might be to create and pass a SPI "Principal" as an argument that can represent both, a user or a role. I can do it in a follow-up PR since this involves more changes.

cawallin replied:
Yeah, I agree this is acceptable (though not ideal). We should be careful about changing the SPI/interfaces of HiveMetastore too much.

cawallin commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/ThriftHiveMetastore.java:
```diff
@@ -769,25 +738,103 @@ public void revokeTablePrivileges(String databaseName, String tableName, String
+    private Set<HivePrivilegeInfo> getRolePrivileges(HivePrincipal hivePrincipal, HiveObjectRef hiveObjectRef)
+    {
+        if (hivePrincipal.getPrincipalType() != ROLE) {
+            throw new PrestoException(HIVE_WRONG_PRINCIPAL_TYPE, "Expected ROLE PrincipalType but found USER");
```
I would expect that you should use a checkState here rather than introducing a new internal Presto error message. This shouldn't be user visible.

amrutagokhale replied:
done

amrutagokhale commented:
@cawallin Addressed your comments. As we discussed, a better solution to handle roles for `GRANT/REVOKE` seems to be qualifying the role name with a `ROLE` keyword to distinguish roles from users. I will do it in a separate PR. As it stands today, `GRANT/REVOKE` code will only support `PUBLIC` role and will always to default to user names in all other cases.

dain commented:
Merged, thanks!
